Ambiq apollo3 serial fix

Wenn0101 opened the pull request:
Major changes:

-made certain parts of the init only happen on the first init
-comment out (is this the best option) part of ambiq hal that "sets up buffer" and disables interrupts if we arent using hal buffer
-change hal to use AM_HAL_UART_FIFO_DISABLE because i couldn't find a threshold that would allow the interrupt to fire on each byte otherwise
-serial_irq_set fixed to enable and disable individual interrupts
-remove fluff, TODO's, and temp code.

outstanding questions
-add FC or remove TODO's?

Wenn0101 committed:
Target-Ambiq apollo3 make serial tests pass

Wenn0101 committed:
Target-Ambiq apollo3 remove unneccessary fluff in serial driver

Wenn0101 committed:
Target-Ambiq apollo3 remove debug code in serial driver

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/sdk/mcu/apollo3/hal/am_hal_uart.c:
```diff
@@ -436,11 +436,11 @@ am_hal_uart_configure(void *pHandle, const am_hal_uart_config_t *psConfig)
-                     psConfig->ui32TxBufferSize,
-                     psConfig->pui8RxBuffer,
-                     psConfig->ui32RxBufferSize);
+    // buffer_configure(pHandle,
+    //                  psConfig->pui8TxBuffer,
+    //                  psConfig->ui32TxBufferSize,
+    //                  psConfig->pui8RxBuffer,
+    //                  psConfig->ui32RxBufferSize);
```
was it necessary to change the HAL here? is it not enough to simply set these 4 buffers as NULL? (i figure the HAL would not do anything with them if they were not provided)

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/sdk/mcu/apollo3/hal/am_hal_uart.c:
```diff
@@ -420,7 +420,7 @@ am_hal_uart_configure(void *pHandle, const am_hal_uart_config_t *psConfig)
                                psConfig->ui32Parity     |
                                psConfig->ui32StopBits   |
-                               AM_HAL_UART_FIFO_ENABLE);
+                               AM_HAL_UART_FIFO_DISABLE);
```
does / should this HAL change go through our AmbiqSuiteSDK mirror branch?

is there really no way to change this setting via HAL calls w/ proper configuration?

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/serial_api.c:
```diff
@@ -126,31 +126,35 @@ void serial_init(serial_t *obj, PinName tx, PinName rx) {
-  // set default baud rate and format
-  serial_baud(obj, 9600);
-  serial_format(obj, 8, ParityNone, 1);
+  if(!obj->serial.uart_control->handle)
```
i see you using the handle as an ```initialized``` flag here... just commenting b/c i have recently doubted whether this method worked for the IOM modules

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/serial_api.c:
```diff
@@ -126,31 +126,35 @@ void serial_init(serial_t *obj, PinName tx, PinName rx) {
-  serial_format(obj, 8, ParityNone, 1);
+  if(!obj->serial.uart_control->handle)
+  {
+    //printf("setting up UART first time?\r\n");
```
prolly don' need this printf... a comment or nothing would suffice

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/serial_api.c:
```diff
@@ -126,31 +126,35 @@ void serial_init(serial_t *obj, PinName tx, PinName rx) {
 
 void serial_free(serial_t *obj)
 {
+  //serialinitialized = false;
```
this variable is no longer used (opting for use of the handle as this flag) so we can remove

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/serial_api.c:
```diff
@@ -225,21 +229,33 @@ void serial_irq_handler(serial_t *obj, uart_irq_handler handler, uint32_t id)
   }
-
-  NVIC_EnableIRQ((IRQn_Type)(UART0_IRQn + obj->serial.uart_control->inst));
+  
```
don't need (style nit)

oclyke commented on targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/serial_api.c:
```diff
@@ -126,31 +126,35 @@ void serial_init(serial_t *obj, PinName tx, PinName rx) {
-  // memset(obj->serial.uart_control->cfg, 0x00, sizeof(am_hal_uart_config_t)); // ensure config begins zeroed
-
-  // config uart pins
+    // config uart pins
```
can we get a line break above this? (style nit)

oclyke commented:
@Wenn0101 just requested some changes mostly b/c I had questions. The only actual changes would be style nits and possibly another way of configuring the HAL to do what we need. Otherwise looks real good

Commit:
style nits fix

removes unused comments, consistent brace style

Commit:
style nit fix

missed one

Commit:
update SDK w/ UART FIFO config option

this adds a new am_hal_uart_configure_fifo function which takes a boolean option whether to enable or disable FIFO operation
